test/rosdep_repo_check/make_rosdep_key.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+
+"""
+Make a rosdep rule from a given an os, release, and package name.
+"""
+
+import argparse
+
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+
+import yaml
+
+
+class Key:
+
+    def __init__(self, *, name=None):
+        # ex: 'ubuntu': ['focal', 'jammy']
+        self._platforms: Dict[str, List[str]] = {}
+        # ex: ('ubuntu', 'focal'): 'cmake'
+        self._packages: Dict[Tuple[str, str], Optional[str]] = {}
+        # Override the name of the rosdep key
+        self._name = name
+
+    def add_rule(
+            self, os: str, release: str, package_name: Optional[str] = None
+        ) -> None:
+        os = str(os)
+        release = str(release)
+        if package_name:
+            package_name = str(package_name)
+        if os not in self._platforms.keys():
+            self._platforms[os] = []
+        if release not in self._platforms[os]:
+            self._platforms[os].append(release)
+        self._packages[(os, release)] = package_name
+
+    def _calculate_name(self) -> Optional[str]:
+        # Prioritize name of ubuntu package as the key name
+        for os in ('ubuntu', 'debian'):
+            if os in self._platforms.keys():
+                for release in self._platforms[os]:
+                    package = self._packages[(os, release)]
+                    if package:
+                        return package
+
+    @property
+    def name(self) -> Optional[str]:
+        if not self._name:
+            self._name = self._calculate_name()
+        return self._name
+
+    def as_dict(self):
+        if self.name is None:
+            raise ValueError('Key has no name')
+        yaml_dict = {}
+
+        for os, releases in self._platforms.items():
+            if os not in yaml_dict:
+                yaml_dict[os] = {}
+            for release in releases:
+                package = self._packages[(os, release)]
+                yaml_dict[os][release] = [package] if package else None
+
+        return {self.name: yaml_dict}
+
+
+def _minimize_rules(yaml_dict, supported_versions):
+    # Remove entries which are entirely null
+    for key_name in list(yaml_dict.keys()):
+        for os in list(yaml_dict[key_name].keys()):
+            os_rules = yaml_dict[key_name][os]
+            if isinstance(os_rules, dict):
+                if any(os_rules.values()):
+                    continue
+            elif isinstance(os_rules, list):
+                if any(os_rules):
+                    continue
+            del yaml_dict[key_name][os]
```

Comment:
I think it got lost in the noise, but we should drop this section because it would remove the explicit null entries you're looking for (in reference to https://github.com/ros/rosdistro/pull/33916#issuecomment-1191981410).

```suggestion
```

Once it's gone, the rest of the minimizing code seems to handle None properly, and will still condense an OS where all versions under it are `null` down to a single `os: null` rule.